builtins.go

```diff
@@ -839,6 +839,59 @@ func builtinPow(i *interpreter, trace traceElement, basev value, expv value) (va
 	return makeDoubleCheck(i, trace, math.Pow(base.value, exp.value))
 }
 
+func builtinSubstr(i *interpreter, trace traceElement, inputStr, inputFrom, inputLen value) (value, error) {
+	strV, err := i.getString(inputStr, trace)
+	if err != nil {
+		msg := fmt.Sprintf("substr first parameter should be a string, got %T", inputStr.getType().name)
+		return nil, makeRuntimeError(msg, i.getCurrentStackTrace(trace))
+	}
+
+	fromV, err := i.getNumber(inputFrom, trace)
+	if err != nil {
+		msg := fmt.Sprintf("substr second parameter should be a number, got %T", inputFrom.getType().name)
+		return nil, makeRuntimeError(msg, i.getCurrentStackTrace(trace))
+	}
+
+	if math.Mod(fromV.value, 1) != 0 {
+		msg := fmt.Sprintf("substr second parameter should be an integer, got %f", fromV.value)
+		return nil, makeRuntimeError(msg, i.getCurrentStackTrace(trace))
+	}
+
+	lenV, err := i.getNumber(inputLen, trace)
+	if err != nil {
+		msg := fmt.Sprintf("substr third parameter should be a number, got %T", inputLen.getType().name)
+		return nil, makeRuntimeError(msg, i.getCurrentStackTrace(trace))
+	}
+
+	if math.Mod(lenV.value, 1) != 0 {
```

Comment:
This could be done slightly more elegantly IMO, with something like math.Trunc(lenV.value) == lenV.value.  Same with fromV before.

Reply:
updated